### PR TITLE
Use locks to support downloading archives in parallel

### PR DESCRIPTION
**Pull request by yurrriq**
Address https://github.com/helm/helm/issues/6140, and supersede https://github.com/helm/helm/pull/4106, by adding lock files to the chart downloader.

I'm not sure how best to test this, but want to be sure to, so I'd love some feedback.

### CODE REVIEW

**Comment by aaronmell** [comment]
> /cc @aaronmell @karuppiah7890

Awesome!

**Comment by yurrriq** [comment]
I've updated this to avoid re-downloading archives and provenance files. Also to lock the provenance file.

**Comment by yurrriq** [comment]
Any feedback? I would love to see this or similar get merged. The concurrency issue is quite a problem for us.

**Comment by karuppiah7890** [comment]
I need to read more to understand the changes. But, it looks like simple and clean changes 😄 Just one suggestion which I always request for - tests. Please do add tests for these. This is to make the software more robust and as much bug free as possible, especially when more features or tweaks are done to this code in the future.

**Comment by karuppiah7890** [comment]
Also, these changes are done for Helm v3, once this is merged, do check if it's possible and easy to make the same changes in v2. Or I could do it. My team uses v2 even now 🙈 Migration will take quite some time 😅

**Comment by yurrriq** [comment]
As I mentioned above, I'm not sure how best to test this and will need some guidance. No need to try to sell me on the merits of testing. :smile: I'm in total agreement.

**Comment by karuppiah7890** [comment]
Ah, cool. My bad I didn't read properly! 🙈  I'll look at the code again and checkout scenarios and comment here what I think and how we can test this in a clean manner 😄

**Comment by yurrriq** [comment]
Rebased.

**Comment by luryus** [comment]
Any chance of getting this reviewed and merged? We constantly run into issues when running multiple Helm upgrades concurrently with helmfile.

**Comment by bacongobbler** [comment]
A couple of quick comments for feedback:

- for testing code paths where two calls running concurrently can run into lock scenarios, consider using [`WaitGroups`](https://gobyexample.com/waitgroups)
- this PR is out of date, and needs another rebase

Given the amount of code being changed and/or modified here, I would not be comfortable merging this PR without writing a few tests to

a) ascertain the base case (#6140)
b) ensure that the fix provided here addresses that case

Thanks!

File: pkg/downloader/chart_downloader.go [hunk 1]
```diff
@@ -114,36 +117,64 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
 		return "", nil, err
 	}
 
-	g, err := c.Getters.ByScheme(u.Scheme)
+	name := filepath.Base(u.Path)
+	destfile := filepath.Join(dest, name)
+
+	// Acquire a file lock for process synchronization
+	fileLock := flock.New(destfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```
Comment (bacongobbler) [inline pkg/downloader/chart_downloader.go:125]: How come this is specificaly set as a 30 second timeout? What made you decide to land on that number as a timeout?

Comment (yurrriq) [reply pkg/downloader/chart_downloader.go:125]: The 30 second timeout is arbitrary, copped from https://github.com/helm/helm/blob/v3.3.1/cmd/helm/repo_add.go#L93

**Comment by yurrriq** [comment]
>     * for testing code paths where two calls running concurrently can run into lock scenarios, consider using [`WaitGroups`](https://gobyexample.com/waitgroups).

Thanks for the tip. I'll have a go sometime in the next few days. Holiday weekend here, so I'm optimistic.

>     * this PR is out of date, and needs another rebase

Yep, will do.

> Given the amount of code being changed and/or modified here, I would not be comfortable merging this PR without writing a few tests to
> 
> a) ascertain the base case (#6140), and ensure the test case fails prior to this patch
> b) ensure that the fix provided here addresses that case

I fully agree. I'll attempt to address both concerns.

**Comment by yurrriq** [comment]
I rebased and wrote tests, and confirmed that #7720 addressed the issue since v3.2.0.

**Pull request closed without merging**